Assignments/ip2/ip2.md

```diff
@@ -0,0 +1,310 @@
+
+{::options parse_block_html="true" /}
+<details><summary markdown="span">View the specification for these methods</summary>
+{% highlight typescript %}
+  {
+  /**
+   * Returns the current state of the board.
+   *
+   * The board is a 3x3 array of TicTacToeCell, which is either 'X', 'O', or undefined.
+   *
+   * The 2-dimensional array is indexed by row and then column, so board[0][0] is the top-left cell,
+   * and board[2][2] is the bottom-right cell
+   */
+  get board(): TicTacToeCell[][]
+  /**
+   * Returns the player with the 'X' game piece, if there is one
+   */
+  get x(): PlayerController | undefined
+  /**
+   * Returns the player with the 'O' game piece, if there is one
+   */
+  get o(): PlayerController | undefined 
+  /**
+   * Returns the number of moves that have been made in the game
+   */
+  get moveCount(): number
+  /**
+   * Returns the winner of the game, if there is one
+   */
+  get winner(): PlayerController | undefined
+  /**
+   * Returns the player whose turn it is, if the game is in progress
+   * Returns undefined if the game is not in progress
+   */
+  get whoseTurn(): PlayerController | undefined
+  /**
+   * Returns true if it is our turn to make a move in the game
+   * Returns false if it is not our turn, or if the game is not in progress
+   */
+  get isOurTurn(): boolean
+  /**
+   * Returns true if the current player is a player in this game
+   */
+  get isPlayer(): boolean
+  /**
+   * Returns the game piece of the current player, if the current player is a player in this game
+   *
+   * Throws an error PLAYER_NOT_IN_GAME_ERROR if the current player is not a player in this game
+   */
+  get gamePiece(): 'X' | 'O'
+  /**
+   * Returns the status of the game.
+   */
+  get status(): GameStatus
+  /**
+   * Returns true if the game is in progress
+   */
+  public isActive(): boolean
+  /**
+   * Updates the internal state of this TicTacToeAreaController to match the new model.
+   *
+   * Calls super._updateFrom, which updates the occupants of this game area and
+   * other common properties (including this._model).
+   *
+   * If the board has changed, emits a 'boardChanged' event with the new board.
+   *
+   * If the turn has changed, emits a 'turnChanged' event with true if it is our turn, and false otherwise.
+   */
+  protected _updateFrom(newModel: GameArea<TicTacToeGameState>): void
+  /**
+   * Sends a request to the server to make a move in the game.
+   * Uses the this._townController.sendInteractableCommand method to send the request.
+   * The request should be of type 'GameMove',
+   * and send the gameID provided by `this._instanceID`.
+   *
+   * If the game is not in progress, throws an error NO_GAME_IN_PROGRESS_ERROR
+   *
+   * @param row Row of the move
+   * @param col Column of the move
+   */
+  public async makeMove(row: TicTacToeGridPosition, col: TicTacToeGridPosition) 
```

Comment:
From my TA:

> I thought that I would need to check if `(!this._instanceID || !this.isActive())` whereas the
> reference solution only checks the `_instanceID`
> 
> The wording of “If the game is not in progress, throws an error NO_GAME_IN_PROGRESS_ERROR”, made me think I need to check `this.isActive()` as well.

Reply:
It was ambiguous on purpose :) I updated my reference solution to check, too. I'll leave the tests and the spec as they are.